src/sagemaker_mxnet_container/training.py

```diff
@@ -13,14 +13,69 @@
 from __future__ import absolute_import
 
 import logging
+import os
+import socket
+import subprocess
+
+from retrying import retry
 
 import sagemaker_containers.beta.framework as framework
 
+ROLES = ['worker', 'scheduler', 'server']
+
 logger = logging.getLogger(__name__)
 
 
+def _scheduler_host(hosts):
+    return hosts[0]
+
+
+def _env_vars_for_role(role, hosts, ps_port, ps_verbose):
+    if role in ROLES:
+        return {
+            'DMLC_NUM_WORKER': str(len(hosts)),
+            'DMLC_NUM_SERVER': str(len(hosts)),
+            'DMLC_ROLE': role,
+            'DMLC_PS_ROOT_URI': _host_lookup(_scheduler_host(hosts)),
+            'DMLC_PS_ROOT_PORT': ps_port,
+            'PS_VERBOSE': ps_verbose,
+        }
+
+    raise ValueError('Unexpected role: {}'.format(role))
+
+
+def _run_mxnet_process(role, hosts, ps_port, ps_verbose):
+    role_env = os.environ.copy()
+    role_env.update(_env_vars_for_role(role, hosts, ps_port, ps_verbose))
+    return subprocess.Popen("python -c 'import mxnet'", shell=True, env=role_env).pid
+
+
+@retry(stop_max_delay=1000 * 60 * 15, wait_exponential_multiplier=100,
+       wait_exponential_max=30000)
+def _host_lookup(host):
+    return socket.gethostbyname(host)
+
+
+def _verify_hosts(hosts):
+    for host in hosts:
+        _host_lookup(host)
+
+
 def train(env):
     logger.info('MXNet training environment: {}'.format(env.to_env_vars()))
```

Comment:
run_module prints the training environment as well. do the customers see those twice?

Reply:
I've seen it only once when looking at integ tests - it seems as though I don't see the SC output unless it fails at that part